src/tracing.c

```diff
@@ -1,43 +1,93 @@
 #include <inttypes.h>
 #include <stdlib.h>
+#include <string.h> /* strstr, strlen */
 #include <time.h>
+#include <unistd.h> /* getpid */
 
+#include "assert.h" /* assert */
 #include "tracing.h"
 
-static inline void noopTracerEmit(struct raft_tracer *t,
-                                  const char *file,
-                                  int line,
-                                  const char *message)
+static inline void noopTracerEmit(struct raft_tracer *t UNUSED,
+                                  const char *file UNUSED,
+                                  unsigned int line UNUSED,
+                                  const char *func UNUSED,
+                                  unsigned int level UNUSED,
+                                  const char *message UNUSED)
 {
-    (void)t;
-    (void)file;
-    (void)line;
-    (void)message;
 }
 struct raft_tracer NoopTracer = {.impl = NULL,
                                  .enabled = false,
                                  .emit = noopTracerEmit};
 
+static inline const char *tracer__short_file_name(const char *fname)
+{
+    static const char top_src_dir[] = "raft/";
+    const char *p;
+
+    p = strstr(fname, top_src_dir);
+    return p != NULL ? p + strlen(top_src_dir) : fname;
+}
+
+static inline const char *tracer__trace_level_name(unsigned int level)
+{
+    static const char *levels[] = {"NONE",   "ALWAYS", "FATAL", "ERROR", "WARN",
+                                   "NOTICE", "INFO",   "DEBUG", "CALL"};
+
+    return level < ARRAY_SIZE(levels) ? levels[level] : levels[0];
+}
+
+static pid_t tracer__pid_cached;
+
+static inline void tracer__emit(const char *file,
+                                unsigned int line,
+                                const char *func,
+                                unsigned int level,
+                                const char *message)
+{
+    struct timespec ts = {0};
+    struct tm tm;
+    /* Use stack pointer instead of explicit thread id */
+    size_t sp = (size_t)&sp;
+
+    clock_gettime(CLOCK_REALTIME, &ts);
+    gmtime_r(&ts.tv_sec, &tm);
+
```

Comment:
some example log line in the comment could be a nice add

Reply:
```
    gmtime_r(&ts.tv_sec, &tm);                                                                                                                                  
                                                                                                                                                                
    /*                                                                                                                                                          
      Example:                                                                                                                                                  
      LIBRAFT[182942] 2023-11-27T14:46:24.912050507 001132 INFO                                                                                                 
      uvClientSend  src/uv_send.c:218 connection available...                                                                                                   
    */                                                                                                                                                          
    fprintf(stderr,                                                                                                                                             
```
done.